spring-ai-core/src/main/java/org/springframework/ai/chat/client/advisor/RetrievalAugmentationAdvisor.java

```diff
@@ -0,0 +1,224 @@
+
+	private final boolean protectFromBlocking;
+
+	private final int order;
+
+	public RetrievalAugmentationAdvisor(DocumentRetriever documentRetriever, @Nullable PromptTemplate promptTemplate,
+			@Nullable Boolean protectFromBlocking, @Nullable Integer order) {
+		Assert.notNull(documentRetriever, "documentRetriever cannot be null");
+		this.documentRetriever = documentRetriever;
+		this.promptTemplate = promptTemplate != null ? promptTemplate : DEFAULT_PROMPT_TEMPLATE;
+		this.protectFromBlocking = protectFromBlocking != null ? protectFromBlocking : false;
+		this.order = order != null ? order : 0;
+	}
+
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	@Override
+	public AdvisedResponse aroundCall(AdvisedRequest advisedRequest, CallAroundAdvisorChain chain) {
+		Assert.notNull(advisedRequest, "advisedRequest cannot be null");
+		Assert.notNull(chain, "chain cannot be null");
+
+		AdvisedRequest processedAdvisedRequest = before(advisedRequest);
+		AdvisedResponse advisedResponse = chain.nextAroundCall(processedAdvisedRequest);
+		return after(advisedResponse);
+	}
+
+	@Override
+	public Flux<AdvisedResponse> aroundStream(AdvisedRequest advisedRequest, StreamAroundAdvisorChain chain) {
+		Assert.notNull(advisedRequest, "advisedRequest cannot be null");
+		Assert.notNull(chain, "chain cannot be null");
+
+		// This can be executed by both blocking and non-blocking Threads
+		// E.g. a command line or Tomcat blocking Thread implementation
+		// or by a WebFlux dispatch in a non-blocking manner.
+		Flux<AdvisedResponse> advisedResponses = (this.protectFromBlocking) ?
+		// @formatter:off
+				Mono.just(advisedRequest)
+						.publishOn(Schedulers.boundedElastic())
+						.map(this::before)
+						.flatMapMany(chain::nextAroundStream)
+				: chain.nextAroundStream(before(advisedRequest));
+		// @formatter:on
+
+		return advisedResponses.map(ar -> {
+			if (onFinishReason().test(ar)) {
+				ar = after(ar);
+			}
+			return ar;
+		});
+	}
+
+	private AdvisedRequest before(AdvisedRequest request) {
+		Map<String, Object> context = new HashMap<>(request.adviseContext());
+
+		// 0. Create a query from the user text and parameters.
+		Query query = new Query(new PromptTemplate(request.userText(), request.userParams()).render());
+
+		// 1. Retrieve similar documents for the original query.
+		List<Document> documents = this.documentRetriever.retrieve(query);
+		context.put(DOCUMENT_CONTEXT, documents);
+
+		// 2. Combine retrieved documents.
+		String documentContext = documents.stream()
+			.map(Content::getContent)
+			.collect(Collectors.joining(System.lineSeparator()));
+
+		// 3. Define augmentation prompt parameters.
+		Map<String, Object> promptParameters = Map.of("query", query.text(), "context", documentContext);
+
+		// 4. Augment user prompt with the context data.
+		UserMessage augmentedUserMessage = (UserMessage) this.promptTemplate.createMessage(promptParameters);
+
+		return AdvisedRequest.from(request)
```

Comment:
We have an issue, #1592 related to how we want to consistently implement builders.  Does `from` go in the 'product' class or in the 'builder' class?